rust/lance-encoding/src/encodings/logical/list.rs

```diff
@@ -528,21 +514,74 @@ impl LogicalPageDecoder for ListPageDecoder {
+            }
+            trace!(
+                "List decoder is waiting for {} rows and {} are already available",
+                num_rows,
+                self.lists_available
+            );
+            if self.lists_available >= num_rows {
+                self.lists_available -= num_rows;
+                return Ok(());
+            }
+            let num_rows = num_rows - self.lists_available;
+            self.lists_available = 0;
+            let offset_wait_start = self.rows_drained + self.lists_available;
+            let item_start = self.offsets[offset_wait_start as usize];
+            let mut items_needed =
+                self.offsets[offset_wait_start as usize + num_rows as usize] - item_start;
+            // First discount any already available items
+            items_needed = items_needed.saturating_sub(
+                self.awaited_item_decoders
+                    .iter()
+                    .map(|d| d.avail() as u64)
+                    .sum::<u64>(),
+            );
+            // The last decoder in the awaited list may not be fully awaited
+            if let Some(last_decoder) = self.awaited_item_decoders.back_mut() {
+                let to_await = items_needed.min(last_decoder.unawaited() as u64) as u32;
+                trace!(
+                    "List decoder waiting for {} items from the last awaited page",
+                    to_await
+                );
+                if to_await > 0 {
+                    last_decoder.wait(to_await, source).await?;
+                    items_needed = items_needed.saturating_sub(last_decoder.avail() as u64);
+                }
+            }
+            // Finally wait for more I/O
+            while items_needed > 0 {
+                let mut next_item_decoder = self.unawaited_item_decoders.pop_front().unwrap();
+                let unawaited = next_item_decoder.unawaited() as u64;
+                trace!(
+                    "List decoder waiting on a new page that has {} items unawaited",
+                    unawaited
+                );
+                let to_await = items_needed.min(unawaited) as u32;
+                // TODO: Seems like this will fail in List<Struct> case
+                next_item_decoder.wait(to_await, source).await?;
+                // Might end up loading more items than needed so use saturating_sub
+                items_needed = items_needed.saturating_sub(next_item_decoder.avail() as u64);
+                self.awaited_item_decoders.push_back(next_item_decoder);
             }
+            // This is technically undercounting a little.  It's possible that we loaded a big items
+            // page with many items and then only needed a few of them for the requested lists.  However,
+            // to find the exact number of lists that are available we would need to walk through the item
+            // lengths and it's faster to just undercount here.
+            self.lists_available += num_rows;
```

Comment:
Not really.  The zeroing of lists_available is because it's in an implicit else...

```
if self.lists_available >= num_rows {
    ...
    return ...
}
// else self.lists_available < num_rows
self.lists_available = 0;
```

We zero it out because it is less typing:

```
// Given self.lists_available < num_rows these statements are equivalent
self.lists_available = self.lists_available.saturating_sub(num_rows);
self.lists_available = 0;
```

The undercounting is a different problem.  For example, the user asked us to load enough data for at least 20 more lists.
We determine that we need 200 items to satisfy this.  We ask the items loader to load 200 items.  The items loader ends up loading 500 items.

Maybe this means we have enough items for 50 lists.  Maybe it means we don't even have enough items for 21 more lists (list 21 could be huge).  We could do the work to figure that out but we don't bother.  If we did that work it would save the user some effort because the next pass through they could say "now I need 30 more lists, oh look! those are already available, I don't need to await".  Instead they say "now I need 30 more lists" and the list decoder then has to do...

```
            // First discount any already available items
            items_needed = items_needed.saturating_sub(
                self.awaited_item_decoders
                    .iter()
                    .map(|d| d.avail() as u64)
                    .sum::<u64>(),
            );
```
to figure out if it already happened to load the items it needs.  I don't think it's really a problem since it just shuffles the work from one spot to another but it might be possible that one approach is more efficient than the other but my guess is it is pretty unlikely it will matter.